torch/distributed/algorithms/_comm_hooks/default_hooks.py

```diff
@@ -80,46 +80,80 @@ def allreduce_hook(state: DefaultState, grad: torch.Tensor):
     and a necessary pre- and post-division of gradients.
 
     Args:
-        state (DefaultState): State information, configures pre- and post-division factors
+        state (DefaultState): State information, configures pre- and post-division factors.
         grad (torch.Tensor): A gradient for the local batch that needs to be communicated across ranks.
     """
+    # Average grad by pre-division factor. Together pre- and post-division factors
+    # lead to an overall averaging by world_size, required for consistency with PyTorch DDP.
+    # This is a two-step process to avoid potential underflow and overflow.
     if state.gradient_predivide_factor > 1:
         grad.div_(state.gradient_predivide_factor)
     dist.all_reduce(grad, group=state.process_group)
+    # Average grad by post-division factor.
     if state.gradient_postdivide_factor > 1:
         grad.div_(state.gradient_postdivide_factor)
 
-def lower_precision_hook(prec: torch.dtype, state: LowPrecisionState, grad: torch.Tensor):
-    grad.data = grad.data.to(prec)
-    allreduce_hook(state, grad)
-    _decompress(state, grad)
+def reduce_scatter_hook(state: DefaultState, grad: torch.Tensor, output: torch.Tensor):
+    r"""
+    This FSDP communication hook implements ``reduce_scatter`` algorithm for
+    sharded FSDP strategies and a necessary pre- and post-division of gradients.
 
-def fp16_compress_hook(state: LowPrecisionState, grad: torch.Tensor):
+    Args:
+        state (DefaultState): State information, configures pre- and post-division factors.
+        grad (torch.Tensor): An unsharded gradient for the local batch that needs to be
+        communicated across ranks.
+        output (torch.Tensor): Stores a single shard of the gradient after ``reduce_scatter``.
+    """
+    # Average grad by pre-division factor.
+    if state.gradient_predivide_factor > 1:
+        grad.div_(state.gradient_predivide_factor)
+    dist._reduce_scatter_base(
+        output, grad, group=state.process_group
+    )
+    # Average grad's shard by post-division factor.
+    if state.gradient_postdivide_factor > 1:
+        output.div_(state.gradient_postdivide_factor)
+
+def _low_precision_hook(prec: torch.dtype, state: LowPrecisionState, grad: torch.Tensor, output: torch.Tensor):
+    grad.data = grad.data.to(prec)
+    if output is not None:
+        output.data = output.data.to(prec)
```

Comment:
I deleted my old comment since it was slightly off.

Corrected: I think we should consider enforcing `output` is the reduce-scatter destination, meaning for `_low_precision_hook`, `output` should already be in the low-precision. This does not affect the fact it will go through `_decompress()` later. This is to save one FP32 -> low precision copy of the sharded gradient tensor.

Reply:
But `output` is not in a reduced precision at this point. `output` is created based on `grad_flatten` before hook is called. `grad_flatten` is being preprocessed inside `FSDP._post_backward_hook`, `output` is also created there, so at that point it is in full precision. `grad_flatten` is converted to reduced precision on L118, so `output` also needs to be converted, there

Reply:
My thought process regarding `reduce_scatter` is: if we don't want an extra conversion FP32-> low_precision, then an `output` should really be created inside of a `hook`, but I don't see how it would be unreachable from `FSDP._post_backward_hook` then

Reply:
> but I don't see how it would be unreachable from FSDP._post_backward_hook then

Can you clarify what this means?

Reply:
If we really cannot avoid this FP32 -> low precision cast here, then that is fine. We can look at it in follow-up PRs.

Reply:
@awgu  Clarification: sorry, it was a typo. "I don't see how it would be reachable from `FSDP._post_backward_hook` then, i.e. since hook returns nothing, we should pass a mutable tensor as an argument to the hook, as a result `output` should be created inside `FSDP._post_backward_hook` and outside of the hook